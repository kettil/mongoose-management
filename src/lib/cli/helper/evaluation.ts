export const mergeEvaluation = <T>(
  mainCall: (item: T | undefined) => T,
  subCalls: ((item: T) => T)[] = [],
  item?: T,
): T => {
  return subCalls.reduce((subItem, evaluation) => evaluation(subItem), mainCall(item));
};
