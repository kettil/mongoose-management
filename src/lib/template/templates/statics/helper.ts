/**
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

import { Schema } from 'mongoose';

import { indexType } from './types';

/**
 *
 * @param schema
 * @param indexes
 */
export const addIndexes = (schema: Schema, indexes: indexType[]) => {
  indexes.forEach(({ fields, options }) => {
    schema.index(fields, options);
  });
};

/**
 *
 * @param schema
 * @param virtual
 */
export const addVirtualProperties = (
  schema: Schema,
  virtual: { [k: string]: { get?: () => void; set?: (val: any) => void } },
) => {
  Object.entries(virtual).forEach(([name, { get, set }]) => {
    const space = schema.virtual(name);

    if (typeof get === 'function') {
      space.get(get);
    }
    if (typeof set === 'function') {
      space.set(set);
    }
  });
};

/**
 *
 * @param name
 * @param items
 */
export const checkDuplicateKeys = (name: string, items: Array<{}>) => {
  const keys = extractKeys(items);
  const duplicates = getDuplicates(keys);

  if (duplicates.length > 0) {
    throw new Error(`Double keys were assigned in the scheme "${name}": ` + duplicates.join(', '));
  }
};

/**
 *
 * @param items
 */
export const getDuplicates = (items: string[]) => {
  const counter: { [k: string]: number } = {};
  const duplicates: string[] = [];

  items.forEach((key) => {
    if (!counter[key]) {
      counter[key] = 1;
    } else {
      counter[key] += 1;
    }
  });

  Object.entries(counter).forEach(([key, count]) => {
    if (count > 1) {
      duplicates.push(key);
    }
  });

  return duplicates;
};

/**
 *
 * @param items
 */
export const extractKeys = (items: Array<{}>) => {
  return items.map((o) => Object.keys(o)).reduce((keys, oKeys) => keys.concat(oKeys), []);
};
