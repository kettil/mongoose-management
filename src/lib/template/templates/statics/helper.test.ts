import { Schema } from 'mongoose';

import * as Helper from './helper';

describe('Check the CollectionDataset class', () => {
  let testSchema: Schema;

  beforeEach(() => {
    testSchema = new Schema({
      foo: String,
      bar: Number,
    });
  });

  test('addIndexes should add the passed index to the schema', () => {
    Helper.addIndexes(testSchema, [
      {
        fields: {
          foo: 1,
        },
        options: {
          unique: true,
        },
      },
    ]);
    expect(testSchema.indexes()).toMatchSnapshot();
  });

  test.each([
    {
      virtualProp: {
        get: () => 'getter',
        set: (val: any) => {
          /* empty */
        },
      },
    },
    {
      virtualProp: {},
    },
  ])('addVirtualProperties should add the passed virtual to the schema', (virtual) => {
    Helper.addVirtualProperties(testSchema, virtual);
    expect(testSchema.indexes()).toMatchSnapshot();
  });

  describe('checkDuplicateKeys in helpers class', () => {
    test('checkDuplicateKeys should not throw if everything is okay', () => {
      const schemaName = 'exampleSchema';
      expect(() => {
        Helper.checkDuplicateKeys(schemaName, [
          {
            foo: true,
            bar: true,
          },
          {
            baz: 1,
            qux: 2,
          },
        ]);
      }).not.toThrow(`Double keys were assigned in the scheme "${schemaName}"`);
    });

    test('checkDuplicateKeys should throw if needed', () => {
      const schemaName = 'exampleSchema';
      expect(() => {
        Helper.checkDuplicateKeys(schemaName, [
          {
            foo: true,
            bar: true,
          },
          {
            baz: 1,
            foo: 2,
          },
        ]);
      }).toThrow(`Double keys were assigned in the scheme "${schemaName}"`);
    });
  });

  test('getDuplicates should return duplicates', () => {
    expect(Helper.getDuplicates(['foo', 'bar', 'baz', 'foo', 'qux', 'baz'])).toEqual(['foo', 'baz']);
  });

  test('extractKeys should return all keys of objects', () => {
    expect(
      Helper.extractKeys([
        {
          foo: true,
          bar: true,
        },
        {
          baz: 1,
          qux: 2,
        },
      ]),
    ).toEqual(['foo', 'bar', 'baz', 'qux']);
  });
});
