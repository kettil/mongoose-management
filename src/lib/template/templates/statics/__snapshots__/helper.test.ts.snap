// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the CollectionDataset class addIndexes should add the passed index to the schema 1`] = `
Array [
  Array [
    Object {
      "foo": 1,
    },
    Object {
      "background": true,
      "unique": true,
    },
  ],
]
`;

exports[`Check the CollectionDataset class addVirtualProperties should add the passed virtual to the schema 1`] = `Array []`;

exports[`Check the CollectionDataset class addVirtualProperties should add the passed virtual to the schema 2`] = `Array []`;
